doc/api/esm.md

```diff
@@ -209,15 +217,53 @@ be interpreted as CommonJS.
 
 The `"main"` field can point to exactly one file, regardless of whether the
 package is referenced via `require` (in a CommonJS context) or `import` (in an
-ES module context). Package authors who want to publish a package to be used in
-both contexts can do so by setting `"main"` to point to the CommonJS entry point
-and informing the package’s users of the path to the ES module entry point. Such
-a package would be accessible like `require('pkg')` and `import
-'pkg/module.mjs'`. Alternatively the package `"main"` could point to the ES
-module entry point and legacy users could be informed of the CommonJS entry
-point path, e.g. `require('pkg/commonjs')`.
-
-## Package Exports
+ES module context).
+
+#### Compatibility with CommonJS-Only Versions of Node.js
+
+Prior to the introduction of Node.js’ support for ES modules, it was a common
+pattern for package authors to include both CommonJS and ES module JavaScript
+sources in their package, with `package.json` `"main"` specifying the CommonJS
+entry point and `package.json` `"module"` specifying the ES module entry point.
+This enabled Node.js to run the CommonJS entry point while build tools such as
+bundlers used the ES module entry point, since Node.js ignored (and still
+ignores) `"module"`.
+
+Node.js can now run ES module entry points, but it remains impossible for a
+package to define separate CommonJS and ES module entry points. This is for good
+reason: the `pkg` variable created from `import pkg from 'pkg'` is not the same
+singleton as the `pkg` variable created from `const pkg = require('pkg')`, so if
+both are referenced within the same app (including dependencies), unexpected
+behavior might occur.
+
+There are two general approaches to addressing this limitation while still
+publishing a package that contains both CommonJS and ES module sources:
+
+1. Document a new ES module entry point that’s not the package `"main"`, e.g.
+   `import pkg from 'pkg/module.mjs'` (or `import 'pkg/esm'`, if using
+   [package exports][]). The package `"main"` would still point to a CommonJS
+   file, and thus the package would remain compatible with older versions of
+   Node.js that lack support for ES modules. For compatibility with other tools,
+   it would be a good idea to additionally define the ES module entry point in
```

Comment:
I don’t think it’s a good idea for node docs to in any way recommend using nonstandard package.json fields.

Reply:
I agree that in general, the Node docs should stick to documenting Node. With `"module"` we have a somewhat special case, as the ecosystem has chosen a de facto standard that Node isn’t joining. Therefore we have to at least mention `"module"`, as it would be confusing to omit it and users should know that Node ignores it; and once you mention that Node doesn’t use it I think it’s important to clarify that package authors _should_ still use it, because supporting the ecosystem beyond Node is still important.

Ultimately we want to provide guidance not just on how older versions of Node can be supported, but on how _users_ of older versions of Node can be supported. Many of those users will use build tools, and so for maximum compatibility a package author should define `"module"`.

Reply:
I don’t agree it’s a de facto standard any more than jsnext:main was. Specifically, a big part of the reason node can never use it *is* community adoption.

It is actively harmful for the community to prematurely standardize on package.json fields, because it disrupts nodes ability to use it - just like polyfilling nonstandard things is harmful to the language.

By defining “module”, an author is cementing likely-infeasible-to-standardize patterns. Node docs should not endorse this.